### PR TITLE
(JP-3029) Update ramp fitting to calculate separate readnoise variance

### PR DESCRIPTION
**Pull request by dmggh**
Resolves [JP-3029](https://jira.stsci.edu/browse/JP-3029)

Closes #

This PR calculates a readnoise variance that can be used by the resampling step, using UNDERSAMP-groups that have been flagged by the undersampling_correction step.   This PR supersedes PR #7443, which will be closed.

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [X] updated or added relevant tests
- [X] updated relevant documentation
- [X] added relevant milestone
- [X] added relevant label(s)
- [X] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Review by kmacdonald-stsci — APPROVED** [review]
Overall looks good, but I think the large variance variables should be imported from STCAL, rather than redefined here.

File: docs/jwst/ramp_fitting/description.rst [hunk 1]
```diff
@@ -317,3 +317,25 @@ Poisson noise, :math:`var^P_{o}`, is stored in the VAR_POISSON extension, the va
 due to read noise, :math:`var^R_{o}`, is stored in the VAR_RNOISE extension, and the
 square-root of the combined variance, :math:`var^C_{o}`, is stored in the ERR
 extension.
+
+
+
+Weighted Readnoise Variance
++++++++++++++++++++++++++++
+If the :ref:`undersampline correction <undersampling_correction_step>`
```
[the remark this answers is not in the document]
Comment (dmggh) [reply docs/jwst/ramp_fitting/description.rst]: Yes it should - thanks.

**Comment by dmggh** [comment]
The regression test link :
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/589/
  
Test failures: 5 for associations, 2 for miri lrs spec

File: jwst/ramp_fitting/ramp_fit_step.py [hunk 2]
```diff
@@ -12,10 +13,15 @@
...
 
 import logging
+import copy
+import warnings
 
 log = logging.getLogger(__name__)
 log.setLevel(logging.DEBUG)
 
+LARGE_VARIANCE = 1.e8
+LARGE_VARIANCE_THRESHOLD = 0.001 * LARGE_VARIANCE
```
[the remark this answers is not in the document]
Comment (dmggh) [reply jwst/ramp_fitting/ramp_fit_step.py]: Good catch. I made that change.

File: jwst/ramp_fitting/ramp_fit_step.py [hunk 3]
```diff
@@ -173,6 +179,202 @@ def create_optional_results_model(input_model, opt_info):
...
+        readnoise only, 3-D float
+
+    var_r4 : ndarray
+        Hypercube of segment- and integration-specific values for the slope
+        variance due to read noise only, 4-D float
+    """
+    nint = groupdq.shape[0]
+    ngroups = groupdq.shape[1]
+    nrows = groupdq.shape[2]
+    ncols = groupdq.shape[3]
```
Comment (hbushouse) [inline jwst/ramp_fitting/ramp_fit_step.py:219]: Could simplify these 4 lines to just `nint, ngroups, nrows, ncols = groupdq.shape`.

File: jwst/ramp_fitting/ramp_fit_step.py [hunk 4]
```diff
@@ -238,16 +440,63 @@ def process(self, input):
...
-                self.weighting, max_cores, dqflags.pixel, suppress_one_group=self.suppress_one_group)
+                input_model, buffsize, self.save_opt, readnoise_2d, gain_2d,
+                self.algorithm, self.weighting, max_cores, dqflags.pixel,
+                suppress_one_group=self.suppress_one_group)
+
+            # Create a gdq to modify if there are undersampling-corrected groups
+            gdq = input_model_W.groupdq.copy()
+
+            # Locate groups where that are flagged with UNDERSAMP
+            wh_undersamp = np.where(np.bitwise_and(gdq.astype(np.int32), dqflags.group['UNDERSAMP']))
```
Comment (hbushouse) [inline jwst/ramp_fitting/ramp_fit_step.py:459]: Any particular reason why type `np.int32` is being used here, as opposed to the normal `np.uint32` for DQ flags?

File: jwst/ramp_fitting/ramp_fit_step.py (hunk 3 above)
Comment (dmggh) [reply jwst/ramp_fitting/ramp_fit_step.py]: Done

File: jwst/ramp_fitting/ramp_fit_step.py [hunk 5]
```diff
@@ -173,6 +179,202 @@ def create_optional_results_model(input_model, opt_info):
...
+            rhi = rlo + nrows
+
+            if rhi > cubeshape[1]:
+                rhi = cubeshape[1]
+
+            gdq_sect = groupdq[num_int, :, rlo:rhi, :]
+            rn_sect = readnoise_2d[rlo:rhi, :]
+            gain_sect = gain_2d[rlo:rhi, :]
+
+            den_r3, num_r3, segs_beg_3, max_seg_i = calc_segs(rn_sect, gdq_sect, group_time)
```
[the remark this answers is not in the document]
Comment (dmggh) [reply jwst/ramp_fitting/ramp_fit_step.py:243]: Done

File: jwst/ramp_fitting/ramp_fit_step.py (hunk 4 above)
Comment (dmggh) [reply jwst/ramp_fitting/ramp_fit_step.py]: No, changed to np.uint32

**Review by hbushouse — COMMENTED** [review]
Looks good. But now that there have been a few small edits to the code it would be good to get another regression test run before we merge. Can you please start one?

**Comment by dmggh** [comment]
Started reg test run 604.

**Comment by hbushouse** [comment]
After updating regtest truth files for other recent mergers, the latest results for this PR branch, in https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/607/, look good. All of the differences are due to updates in master that are not included in this branch.

**Review by hbushouse — APPROVED** [review]

**Pull request merged**